### PR TITLE
Remove Swift dependency on non-public C++ APIs

### PR DESCRIPTION
**Pull request by externl**
This PR removes Swift's dependency on non public C++ APIs. 

Namely:
- We don't call C++'s `traceSlicing`. This means we'll log every call now in Swift. Currently C++ caches ids and only logs once per id, which seems a bit excessive to me.
- Remove calls to `escapeString`. We only use this to escape facets and category names when throwing a few exceptions. Not having this is not 100% correct. We should either copy the logic into Swift or make the C++ function public (or just not worry about it) in a future PR. I'm not sure how we handle this in Python.
- Remove  `errorToString` and `errorToStringDNS` which were just wrappers of their `IceInternal` counterpart. They were untested and unused.

### CODE REVIEW

File: swift/src/Ice/InputStream.swift [hunk 1]
```diff
@@ -361,11 +361,8 @@ public class InputStream {
             return
         }
 
-        ICETraceUtil.traceSlicing(
-            kind: sliceType == SliceType.ExceptionSlice ? "exception" : "object",
-            typeId: typeId,
-            slicingCat: "Slicing",
-            logger: LoggerWrapper(handle: communicator.getLogger()))
+        let kind = sliceType == SliceType.ExceptionSlice ? "exception" : "object"
+        communicator.getLogger().trace(category: "Slicing", message: "unknown \(kind) type '\(typeId)'")
```
Comment (pepone) [inline swift/src/Ice/InputStream.swift:365]: Should we apply this change to other mappings?

**Review by pepone — APPROVED** [review]

**Comment by bernardnormier** [comment]
I don't understand the motivation for this PR. What's wrong with depending on symbols defined in IceImpl?

We have many ICE_API symbols in IceImpl (e.g. used by the generated code). And there many undocumented APIs in namespace Ice (say ConnectionI). For me, the distinction between namespace Ice and IceImpl is fairly meaningless.

Then:
> We don't call C++'s traceSlicing. This means we'll log every call now in Swift. Currently C++ caches ids and only logs once per id, which seems a bit excessive to me.

This is a rather unfortunate change. In all langue mappings, the have this "duplicate log suppression" logic for class/exception slicing diagnostics. But now, you want to change this logic for Swift and embrace duplicate and very verbose tracing?

> Remove calls to escapeString. We only use this to escape facets and category names when throwing a few exceptions

I would rather keep this escapeString as it's more correct and consistent with what we do in other language mappings.

**Comment by externl** [comment]
> I don't understand the motivation for this PR. What's wrong with depending on symbols defined in IceImpl?

These are not part of the public headers, so they're not part of the public API and will not be in the framework.

> We have many ICE_API symbols in IceImpl (e.g. used by the generated code). And there many undocumented APIs in namespace Ice (say ConnectionI). For me, the distinction between namespace Ice and IceImpl is fairly meaningless.
> 
> Then:
> 
> > We don't call C++'s traceSlicing. This means we'll log every call now in Swift. Currently C++ caches ids and only logs once per id, which seems a bit excessive to me.
> 
> This is a rather unfortunate change. In all langue mappings, the have this "duplicate log suppression" logic for class/exception slicing diagnostics. But now, you want to change this logic for Swift and embrace duplicate and very verbose tracing?

I'm also fine applying the same logic in Swift or making the C++ API public. 

> > Remove calls to escapeString. We only use this to escape facets and category names when throwing a few exceptions
> 
> I would rather keep this escapeString as it's more correct and consistent with what we do in other language mappings.

I agree, we'll need to either make the API public or copy the logic into Swift.

**Comment by bernardnormier** [comment]
For APIs that we use (traceSlicing, espaceString) from Swift (and presumably other C++-based language mappings), I would move them to "public" headers in include/Ice.

Even though they are in public headers, they remain undocumented and in namespace IceImpl.

**Comment by externl** [comment]
Swift is actually the only language to use them.

**Comment by externl** [comment]
Looking a little more it's actually a bit worse. I think I can fix `escapeString` to not use instance and just get the toStringMode from properties. 

However we're currently Instance in Communicator to get the defaultAndOverrides for a few things. We use these to create our own Swift version of DefaultsAndOverrides that we pass into OutputStream.

**Comment by externl** [comment]
Closing this for now as there's no simple way to not rely all of the internal headers. It's fine that they're not in the XCFramework as we always do a build from the repo were the private headers are available.

**Pull request closed without merging**